### ISSUE TITLE
Constrasrive learning explained.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#50]
- Vishal Rajput. [Constrasrive learning explained](https://medium.com/aiguys/contrastive-learning-explained-17fa79f707bf).
 - Roushanak Rahmat. [Contrastive Learning TensorFlow Tutorial](https://medium.com/@roushanakrahmat/contrastive-learning-tensorflow-tutorial-c114ff964a48).

**Comment by NorbertZheng** [comment]
## Overview
We all know that
- getting hands-on clean and labeled data is extremely rare in the real world.

Researchers have been trying to develop methods that work with partially labeled data for years. There are quite a few techniques in Semi-supervised learning that works quite decently with partially labeled data. But still, most of them suffer significantly in the case of Deep learning. In this blog post, we are going to discuss a strategy that doesn’t require any labels, and it's called **Contrastive Learning**. So, without further ado let’s dive deep into the concept of contrastive learning.

 - **Semi-Supervised Learning**: A small amount of data that is correctly labeled data is used with large unlabeled data. **The model makes predictions on the unlabeled data and where it is very sure, those samples are added to the next iteration of model training.** It’s an iterative process in which models keep getting better with more and more trained data.
 - **Self-Supervised Learning**: In the self-supervised learning technique, the model depends on the underlying structure of data to predict outcomes. It involves no labeled data.

Note: 80% of the time spent in a supervised learning ML project is invested in acquiring and cleaning the data for model training.

**Comment by NorbertZheng** [comment]
## What is Contrastive Learning?
Contrastive Learning is a technique that is used generally in vision tasks lacking labeled data. By using the principle of contrasting samples against each other it learns
 - attributes that are common between data classes,
 - attributes that set apart a data class from another.

As the name suggests, samples are contrasted against each other, and those belonging to the same distribution or class are pulled together in the embedding space. In contrast, those belonging to different distributions are pushed against each other.

Therefore, contrastive learning is generally considered to be a form of **self-supervised learning**, because it does not require labeled data from external sources in order to train the model to predict the difference or relationship between two input items. It is often used for representation learning, where the goal is to learn useful and meaningful representations of the input data.

![image](https://user-images.githubusercontent.com/37235023/215933496-b68a9bdc-332f-4908-9526-f7a2a43cfd95.png)
Image from [Contrastive Self-Supervised Learning | Ankesh Anand](https://ankeshanand.com/blog/2020/01/26/contrative-self-supervised-learning.html).

In contrastive learning, the model is presented with pairs of items and is trained to predict whether the two items are related or not. For example, the model might be presented with pairs of images and asked to predict whether the images are of the same object or not. The model is then trained to minimize the error in its predictions by adjusting its internal representations of the input data.

 - Contrastive learning can be used to learn pixel-level features from images that are similar to the way that humans process visual information.

In particular, contrastive learning can be used to learn features that are invariant to certain transformations, such as translation or rotation, which are important for recognizing objects in natural images.

**Comment by NorbertZheng** [comment]
## Similarity/Difference measurement
![image](https://user-images.githubusercontent.com/37235023/215934326-5f4f60fd-c9b3-40c7-9db7-ae4ebbb1643d.png)
 - $x+$ is data point similar to $x$, referred to as a positive sample.
 - $x-$ is a data point dissimilar to $x$, referred to as a negative sample.

**Comment by NorbertZheng** [comment]
## Contrastive learning frameworks
### SimCLR
This model is developed by Google Brain, it is a framework for contrastive learning of visual representations. Its basic working principle is to
 - maximize the agreement between different augmented versions of the same sample using a contrastive loss in the latent space.

The framework of the SimCLR method is shown below.

![image](https://user-images.githubusercontent.com/37235023/215936160-915d59c3-69f2-4d70-a2b9-c547cea1b122.png)
A simple framework for contrastive learning of visual representations. Two separate data augmentation operators are sampled from the same family of augmentations ($t \sim \mathcal{T}$ and $t' \sim \mathcal{T}$) and applied to each data example to obtain two correlated views. A base encoder network $f(\cdot)$ and a projection head $g(\cdot)$ are trained to maximize agreement using a contrastive loss. After training is completed, we throw away the projection head $g(\cdot)$ and use encoder $f(\cdot)$ and representation $h$ for downstream tasks.

A nicer illustration is as follows:
![image](https://user-images.githubusercontent.com/37235023/215936896-ce1a798a-7c16-4ae5-8cd4-289bbc2c4075.png)

**Comment by NorbertZheng** [comment]
### Modules within SimCLR
 - **Data augmentation module**: Transforms a given data sample (image) randomly to create two views of the same example ($x_{i}$ and $x_{j}$ in the diagram above). These represent the positive pairs. The SimCLR framework applies the following three augmentations:
    - random crop and resizing (with random flip),
    - color distortions,
    - Gaussian blur.

   According to the results obtained by the authors, random cropping and color distortion are essential for achieving good performance.
 - **Neural network-based encoder**: Denoted by $f(\cdot)$ in the architecture diagram, to extract representative vectors from the augmented data samples. Although we can use any network as the backbone encoder, the authors chose the ResNet model for simplicity. The features are extracted after the final averaging pooling layer of the ResNet-50 model.
 - **Small neural network projection head**: Denoted by $g(\cdot)$ is incorporated to map the extracted representative vectors to a common latent space. This will allow the contrastive loss to be implemented. The authors use a simple Multi-Layer Perceptron with one hidden layer for this purpose. They investigated and found that applying the contrastive loss in this latent space produces better results than directly evaluating the loss over the features extracted from ResNet-50.
 - **Contrastive loss function**: Contrastive loss is defined as the negative log-likelihood of the correct data point, given the target data point and the two data points being compared.

**Comment by NorbertZheng** [comment]
### Contrastive loss function
The loss function used here is called Normalized Temperature-scaled Cross-Entropy or NT-Xent loss. It is a modification of the multi-class $N$-pair loss with an addition of the temperature ( $T$ ) parameter. In the multi-class N-pair loss sampling,
 - instead of sampling a single negative sample, an $N$ number of negative samples are sampled along with one anchor and one positive sample.

![image](https://user-images.githubusercontent.com/37235023/215938484-8f4f9003-2791-4d10-8289-757cb6bb6632.png)
Image from [The Illustrated SimCLR Framework (amitness.com)](https://amitness.com/2020/03/illustrated-simclr/).

It’s worth noting that NT-Xent is not directly related to the **cosine transform**. The cosine transform is often used in contrastive learning, a machine learning technique for training models to recognize similarities and differences between pairs of data points, but it is not a component of the NT-Xent loss function.

![image](https://user-images.githubusercontent.com/37235023/215939060-2648cc1a-ee43-46c6-9196-8e303d722289.png)
Image from [The Illustrated SimCLR Framework (amitness.com)](https://amitness.com/2020/03/illustrated-simclr/).

![image](https://user-images.githubusercontent.com/37235023/215939138-42143c2d-7572-4bf6-b3da-b03be20b5744.png)
Image from [The Illustrated SimCLR Framework (amitness.com)](https://amitness.com/2020/03/illustrated-simclr/).

**Comment by NorbertZheng** [comment]
## Contrastive Learning Tensorflow Tutorial
Let’s implement the contrastive learning to learn pixel-level features from the cifar10 dataset of images using TensorFlow:

```python
import tensorflow as tf

# Load the dataset of images
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

# Preprocess the data
x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

# Define the model architecture
model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, 3, activation='relu', input_shape=(32, 32, 3)),
    tf.keras.layers.MaxPooling2D(),
    tf.keras.layers.Conv2D(64, 3, activation='relu'),
    tf.keras.layers.MaxPooling2D(),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

# Define the contrastive loss function
def contrastive_loss(y_true, y_pred):
    margin = 1
    return tf.reduce_mean(y_true * tf.square(y_pred) +
                         (1 - y_true) * tf.square(tf.maximum(margin - y_pred, 0)))

# Compile the model with the contrastive loss function
model.compile(optimizer='adam', loss=contrastive_loss, metrics=['accuracy'])

# Define a function to generate pairs of images for training
def generate_pairs(x, y):
  while True:
    indices = np.random.permutation(len(x))
    for i in range(0, len(x), 2):
      a = x[indices[i]]
      b = x[indices[i+1]]
      yield ([a, b], [y[indices[i]], y[indices[i+1]]])

# Use the `fit` method to train the model on the generated pairs of images
model.fit(generate_pairs(x_train, y_train), epochs=5,
          validation_data=generate_pairs(x_test, y_test))
```

**Comment by NorbertZheng** [comment]
## Contrastive Learning Tensorflow Tutorial
Let’s implement the contrastive learning to learn pixel-level features from the cifar10 dataset of images using TensorFlow:

```python
import tensorflow as tf

# Load the dataset of images
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

# Preprocess the data
x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

# Define the model architecture
model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, 3, activation='relu', input_shape=(32, 32, 3)),
    tf.keras.layers.MaxPooling2D(),
    tf.keras.layers.Conv2D(64, 3, activation='relu'),
    tf.keras.layers.MaxPooling2D(),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.Dense(1, activation='sigmoid')
])

# Define the contrastive loss function
def contrastive_loss(y_true, y_pred):
    margin = 1
    return tf.reduce_mean(y_true * tf.square(y_pred) +
                         (1 - y_true) * tf.square(tf.maximum(margin - y_pred, 0)))

# Compile the model with the contrastive loss function
model.compile(optimizer='adam', loss=contrastive_loss, metrics=['accuracy'])

# Define a function to generate pairs of images for training
def generate_pairs(x, y):
  while True:
    indices = np.random.permutation(len(x))
    for i in range(0, len(x), 2):
      a = x[indices[i]]
      b = x[indices[i+1]]
      yield ([a, b], [y[indices[i]], y[indices[i+1]]])

# Use the `fit` method to train the model on the generated pairs of images
model.fit(generate_pairs(x_train, y_train), epochs=5,
          validation_data=generate_pairs(x_test, y_test))
```

**Comment by NorbertZheng** [comment]
## References
1. [The Illustrated SimCLR Framework (amitness.com)](https://amitness.com/2020/03/illustrated-simclr/).
2. [Review on Self-Supervised Contrastive Learning | by Lilit Yolyan | Towards Data Science](https://towardsdatascience.com/review-on-self-supervised-contrastive-learning-93171f695140).
3. [Supervised Contrastive Learning (keras.io)](https://keras.io/examples/vision/supervised-contrastive-learning/).
